packages/contracts/contracts/Unirep.sol

```diff
@@ -408,25 +453,19 @@ contract Unirep is IUnirep, zkSNARKHelper, VerifySignature {
         );
         uint256 maxReputationBudget = config.maxReputationBudget;
         verifyProofNullifier(proofNullifier);
-        verifyAttesterSignUp(msg.sender);
-        verifyAttesterIndex(msg.sender, publicSignals[maxReputationBudget + 3]);
-        verifyAttesterFee();
 
         if (publicSignals[maxReputationBudget + 2] != currentEpoch)
             revert EpochNotMatch();
-        if (attesters[msg.sender] != publicSignals[maxReputationBudget + 3])
-            revert AttesterIdNotMatch(publicSignals[maxReputationBudget + 3]);
-
-        if (publicSignals[0] > maxEpochKey) revert InvalidEpochKey();
-
-        // Add to the cumulated attesting fee
-        collectedAttestingFee = collectedAttestingFee.add(msg.value);
 
         // attestation of spending reputation
         Attestation memory attestation;
-        attestation.attesterId = attesters[msg.sender];
+        attestation.attesterId = publicSignals[maxReputationBudget + 3];
         attestation.negRep = publicSignals[maxReputationBudget + 4];
 
+        assertValidAttestation(msg.sender, attestation, publicSignals[0]);
```

Comment:
also here
`attesters[msg.sender] == maxReputationBudget + 3`

Reply:
Actually this is verified by [this function](https://github.com/Unirep/Unirep/pull/84/files/bbba5342b873497fb3e090d28bc7b90f857594ee#diff-bd4415896551787d4d50aa8e8bf1cf39f56c05251c9f8fdb9b1746aca509ec02R208).